18F/identity-idp
# Replace Webpack dev server with zero-dependency alternative

## 🛠 Summary of changes

Implements a lightweight, zero-dependency alternative to `webpack-dev-server`.

**Why?**

`webpack-dev-server` is pretty feature rich, but we use very few of its features. Arguably, we could get rid of it altogether and everything would be fine, but some features like pausing the request while compilation is in-progress are useful to improve confidence during development when testing changes.

Of note is that `webpack-dev-server` creates a very deep dependency tree because it's implemented as an [Express](https://www.npmjs.com/package/express)-based web server. Moving away from it has additional benefits:

- **Decreases risk and maintenance burden for security advisories.** Express and related dependencies are frequent targets for security advisories, and consequently these security advisories bubble up through `webpack-dev-server`. All 9 of the [9 open security advisories](https://github.com/18F/identity-idp/security/dependabot) trace back to `webpack-dev-server`.
- **Decreases inherent risk and installation time for third-party dependencies.** Dropping `webpack-dev-server` drops 146 dependencies (!!!)†.

† _Counted via: `git show 7ac0f1baeb -- yarn.lock | grep -E "^-\S" | wc -l`_

## 📜 Testing Plan

Verify that JavaScript continues to compile and behave as expected in local development:

1. `rm -rf public/packs`
2. `make run`
3. Go to http://localhost:3000
4. Verify that JavaScript behaviors still work (e.g. "Here's how you know" banner collapse, language selector, field validation, etc).
